### PR TITLE
Enable SwiftCompilerSources on Windows

### PR DESCRIPTION
**Pull request by eeckstein**
With some hacks and workarounds I could get SwiftCompilerSources working on my local Windows machine.
To-do: also enable it in lldb (see my comment in the top-level CMakeLists file).

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci test windows

File: CMakeLists.txt [hunk 1]
```diff
@@ -730,6 +730,17 @@ option(SWIFT_ENABLE_BACKTRACING
 set(SWIFT_DARWIN_DEPLOYMENT_VERSION_MACCATALYST "14.5" CACHE STRING
   "Minimum deployment target version for macCatalyst")
 
+# A tempoarary hack: force enabling HOSTTOOLS mode on Windows.
+# Right now, SwiftCompilerSources cannot be enabled for lldb because the
+# Windows build script builds lldb before swift and that fails with a missing
+# dependency on swiftrt (which is built within swift).
```
Comment (compnerd) [inline CMakeLists.txt:736]: lldb is not built before swift, it is built alongside as a single unified build. This implies that there is a missing dependency in the LLDB build definition.

File: SwiftCompilerSources/CMakeLists.txt [hunk 2]
```diff
@@ -106,6 +106,13 @@ function(add_swift_compiler_modules_library name)
       "-Xcc" "-std=c++17"
       "-Xcc" "-DCOMPILED_WITH_SWIFT" "-Xcc" "-DSWIFT_TARGET"
       "-Xcc" "-UIBOutlet" "-Xcc" "-UIBAction" "-Xcc" "-UIBInspectable")
+  
+  if(CMAKE_SYSTEM_NAME STREQUAL "Windows")
+    list(APPEND swift_compile_options "-static")
+    # Workaround a crash in the LoadableByAddress pass
```
Comment (compnerd) [inline SwiftCompilerSources/CMakeLists.txt:112]: Is there a radar and Swift issue that tracks this?

File: SwiftCompilerSources/CMakeLists.txt [hunk 3]
```diff
@@ -106,6 +106,13 @@ function(add_swift_compiler_modules_library name)
       "-Xcc" "-std=c++17"
       "-Xcc" "-DCOMPILED_WITH_SWIFT" "-Xcc" "-DSWIFT_TARGET"
       "-Xcc" "-UIBOutlet" "-Xcc" "-UIBAction" "-Xcc" "-UIBInspectable")
+  
+  if(CMAKE_SYSTEM_NAME STREQUAL "Windows")
+    list(APPEND swift_compile_options "-static")
```
Comment (compnerd) [inline SwiftCompilerSources/CMakeLists.txt:111]: Statically linking the standard library is not supported - how do we ensure that this is not applied to it?

File: SwiftCompilerSources/Sources/Basic/Utils.swift [hunk 4]
```diff
@@ -52,7 +52,9 @@ public extension NoReflectionChildren {
   var customMirror: Mirror { Mirror(self, children: []) }
 }
 
+#if !os(Windows)
 public var standardError = CFileStream(fp: stderr)
+#endif
```
Comment (compnerd) [inline SwiftCompilerSources/Sources/Basic/Utils.swift:57]: You should be able to implement this with the Win32 + ucrt APIs to convert the stderr handle to a `FILE *`.

File: SwiftCompilerSources/Sources/Optimizer/InstructionSimplification/SimplifyConvertEscapeToNoEscape.swift [hunk 5]
```diff
@@ -28,6 +28,9 @@ private extension ConvertEscapeToNoEscapeInst {
   ///   %3 = thin_to_thick_function %1 to $@noescape () -> ()
 
   func tryCombineWithThinToThickOperand(_ context: SimplifyContext) {
+    // compiling bridged.getFunctionTypeWithNoEscape crashes the 5.10 Windows compiler
```
Comment (compnerd) [inline SwiftCompilerSources/Sources/Optimizer/InstructionSimplification/SimplifyConvertEscapeToNoEscape.swift:31]: If this works on newer compilers, perhaps we should also include a version check?

File: CMakeLists.txt (hunk 1 above)
Comment (eeckstein) [reply CMakeLists.txt]: Thanks for clarifying. I updated the comment (We can fix the underlying problem in a follow-up PR).

File: SwiftCompilerSources/CMakeLists.txt (hunk 2 above)
Comment (eeckstein) [reply SwiftCompilerSources/CMakeLists.txt]: Yes, I just created github issues and added the links in comments

**Comment by eeckstein** [comment]
@swift-ci test Windows

File: SwiftCompilerSources/CMakeLists.txt (hunk 3 above)
Comment (compnerd) [reply SwiftCompilerSources/CMakeLists.txt]: Ah, okay, this flag is _only_ applied to `SwiftCompilerSources`? If so, that should be fine. I just wanted to ensure that it didn't accidentally leak. You are correct that it only applies to the module that you are building.

**Comment by eeckstein** [comment]
@swift-ci test Windows

**Comment by eeckstein** [comment]
@swift-ci test Windows

**Comment by eeckstein** [comment]
@swift-ci test windows

**Comment by eeckstein** [comment]
The build is failing with
```
FAILED: tools/swift/SwiftCompilerSources/Basic.o T:/5/tools/swift/SwiftCompilerSources/Basic.o 
cmd.exe /C "cd /D C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\SwiftCompilerSources && T:\toolchains\swift-5.10-DEVELOPMENT-SNAPSHOT-2024-01-18-a-windows10\LocalApp\Programs\Swift\Toolchains\0.0.0+Asserts\usr\bin\swiftc.exe -c -o T:/5/tools/swift/SwiftCompilerSources/Basic.o -resource-dir T:/toolchains/swift-5.10-DEVELOPMENT-SNAPSHOT-2024-01-18-a-windows10/LocalApp/Programs/Swift/Platforms/0.0.0/Windows.platform/Developer/SDKs/Windows.sdk/usr/lib/swift -target x86_64-unknown-windows-msvc -module-name Basic -emit-module -emit-module-path T:/5/tools/swift/SwiftCompilerSources/Basic.swiftmodule -parse-as-library C:/Users/swift-ci/jenkins/workspace/swift-PR-windows/swift/SwiftCompilerSources/Sources/Basic/SourceLoc.swift C:/Users/swift-ci/jenkins/workspace/swift-PR-windows/swift/SwiftCompilerSources/Sources/Basic/StringParser.swift C:/Users/swift-ci/jenkins/workspace/swift-PR-windows/swift/SwiftCompilerSources/Sources/Basic/Utils.swift -wmo -color-diagnostics -Xfrontend -validate-tbd-against-ir=none -Xfrontend -disable-target-os-checking -Xcc -std=c++17 -Xcc -DCOMPILED_WITH_SWIFT -Xcc -DSWIFT_TARGET -Xcc -UIBOutlet -Xcc -UIBAction -Xcc -UIBInspectable -cxx-interoperability-mode=default -O -cross-module-optimization -Xcc -UNDEBUG -Xcc -DPURE_BRIDGING_MODE -Xfrontend -disable-legacy-type-info -static -Xllvm -sil-disable-pass=loadable-address -Xcc -I -Xcc C:/Users/swift-ci/jenkins/workspace/swift-PR-windows/llvm-project/llvm/include -Xcc -I -Xcc T:/5/include -Xcc -I -Xcc C:/Users/swift-ci/jenkins/workspace/swift-PR-windows/llvm-project/llvm/../clang/include -Xcc -I -Xcc T:/5/tools/clang/include -Xcc -I -Xcc C:/Users/swift-ci/jenkins/workspace/swift-PR-windows/swift/include -Xcc -I -Xcc T:/5/tools/swift/SwiftCompilerSources/../include -I T:/5/tools/swift/SwiftCompilerSources"
<unknown>:0: error: cannot load underlying module for 'CxxStdlib'
```
 I didn't see this problem when building locally on my machine.

**Comment by eeckstein** [comment]
@swift-ci test Windows

**Comment by eeckstein** [comment]
@swift-ci test Windows

**Comment by compnerd** [comment]
> The build is failing with
> 
> ```
> FAILED: tools/swift/SwiftCompilerSources/Basic.o T:/5/tools/swift/SwiftCompilerSources/Basic.o 
> cmd.exe /C "cd /D C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\SwiftCompilerSources && T:\toolchains\swift-5.10-DEVELOPMENT-SNAPSHOT-2024-01-18-a-windows10\LocalApp\Programs\Swift\Toolchains\0.0.0+Asserts\usr\bin\swiftc.exe -c -o T:/5/tools/swift/SwiftCompilerSources/Basic.o -resource-dir T:/toolchains/swift-5.10-DEVELOPMENT-SNAPSHOT-2024-01-18-a-windows10/LocalApp/Programs/Swift/Platforms/0.0.0/Windows.platform/Developer/SDKs/Windows.sdk/usr/lib/swift -target x86_64-unknown-windows-msvc -module-name Basic -emit-module -emit-module-path T:/5/tools/swift/SwiftCompilerSources/Basic.swiftmodule -parse-as-library C:/Users/swift-ci/jenkins/workspace/swift-PR-windows/swift/SwiftCompilerSources/Sources/Basic/SourceLoc.swift C:/Users/swift-ci/jenkins/workspace/swift-PR-windows/swift/SwiftCompilerSources/Sources/Basic/StringParser.swift C:/Users/swift-ci/jenkins/workspace/swift-PR-windows/swift/SwiftCompilerSources/Sources/Basic/Utils.swift -wmo -color-diagnostics -Xfrontend -validate-tbd-against-ir=none -Xfrontend -disable-target-os-checking -Xcc -std=c++17 -Xcc -DCOMPILED_WITH_SWIFT -Xcc -DSWIFT_TARGET -Xcc -UIBOutlet -Xcc -UIBAction -Xcc -UIBInspectable -cxx-interoperability-mode=default -O -cross-module-optimization -Xcc -UNDEBUG -Xcc -DPURE_BRIDGING_MODE -Xfrontend -disable-legacy-type-info -static -Xllvm -sil-disable-pass=loadable-address -Xcc -I -Xcc C:/Users/swift-ci/jenkins/workspace/swift-PR-windows/llvm-project/llvm/include -Xcc -I -Xcc T:/5/include -Xcc -I -Xcc C:/Users/swift-ci/jenkins/workspace/swift-PR-windows/llvm-project/llvm/../clang/include -Xcc -I -Xcc T:/5/tools/clang/include -Xcc -I -Xcc C:/Users/swift-ci/jenkins/workspace/swift-PR-windows/swift/include -Xcc -I -Xcc T:/5/tools/swift/SwiftCompilerSources/../include -I T:/5/tools/swift/SwiftCompilerSources"
> <unknown>:0: error: cannot load underlying module for 'CxxStdlib'
> ```
> 
> I didn't see this problem when building locally on my machine.

This is likely another missing dependency in the build graph (CC: @etcwilde). The build is shuffled on differently on different hosts. The dependencies should be accurate to avoid these types of spurious failures.

**Comment by eeckstein** [comment]
> This is likely another missing dependency in the build graph 

I already fixed it. I was using some wrong build options

**Comment by eeckstein** [comment]
@swift-ci test

**Pull request merged**